### ISSUE TITLE
Allow to call getter/setter logic When subclassing

### ISSUE THREAD

**Issue by mwisnicki** [Microsoft/TypeScript#338]
http://typescript.codeplex.com/workitem/645

At the moment it's not possible to call super getter/setter when overriding properties:

``` typescript
class A {
    private _property: string;
    get property() { return this._property; }
    set property(value: string) { this._property = value }
}

class B extends A {
    set property(value: string) {
        super.property = value + " addition";
    }
}
```

**Comment by mtraynham** [comment]
Has anyone considered, just copying the super implementation to the child class and giving it a different variable name?  If your class overrides the property, provide one "private" implementation to get the super property and one to get the overridden subclass property?

```
class A {
    public id: string;
    private _name: string;

    public get name(): string {
        return this.name;
    }
    public set name(name: string) {
        this._name = name;
    }
}

class B extends A {
    public get id(): string {
        return super.id + 'foo';
    }
    public set id(id: string) {
        super.id = id;
    }

    public get name(): string {
        return super.name + 'bar';
    }
    public set name(name: string) {
        super.name = name;
    }
}
```

Class B Becomes:
```
function B(_super) {
    Object.defineProperty(B.prototype, "id", {
        get: function () {
            return this._id + 'bar';
        },
        set: function (id) {
            this._id = id;
        },
        enumerable: true,
        configurable: true
    });

   Object.defineProperty(B.prototype, "_name", {
        get: function () {
            // super implementation
            return __name;
        },
        set: function (name) {
            // super implementation
            this.__name = name;
        },
        enumerable: false,
        configurable: false
    });

    Object.defineProperty(B.prototype, "name", {
        get: function () {
            return this._name + 'bar';
        },
        set: function (name) {
            this._name = name;
        },
        enumerable: true,
        configurable: true
    });
}
```
To address @RyanCavanaugh 's point, you could just move `y` on the subclass to `_y`.  As long as the variable name doesn't conflict with something previously defined by the user.

**Comment by RyanCavanaugh** [comment]
It's not at all guaranteed that you have access to the base class's implementation code.

**Comment by mtraynham** [comment]
Well there is `Object.getOwnPropertyDescriptor(A.prototype, 'name')` which returns the implementation details for A's `name` implementation, albeit it is more of a run-time feature rather than compile time.

**Comment by mwisnicki** [comment]
So what's wrong with `get_Foo`/`set_Foo` + `Object.defineProperty`, where `super.foo` gets rewritten to setter/getter while any other reference is left as is ?

**Comment by jiaweihli** [comment]
The workarounds seem to boil down to:

  - **[application maintenance issue]** duplicating getter/setter implementations from parent classes
  - **[performance issue, non-idiomatic]** using java-style `setXYZ / getXYZ` methods instead of native ES5 methods
  - **[ts compiler, application maintenance issues]** using a reflective `superGet` method & relying on internal typescript compiler semantics (and `eval`, since `super` compiles to `_super.prototype.` - note trailing period), e.g.

```typescript
var getPropertyDescriptor = function(subject, name) {
  var pd = Object.getOwnPropertyDescriptor(subject, name);
  var proto = Object.getPrototypeOf(subject);
  while (pd === undefined && proto !== null) {
    pd = Object.getOwnPropertyDescriptor(proto, name);
    proto = Object.getPrototypeOf(proto);
  }
  return pd;
}

class Test {
    get foo(): number { return 5; }
}

class Test2 extends Test { }

class Test3 extends Test2 {
    private superGet(property) {
        return getPropertyDescriptor(eval('_super.prototype'), property).get();
     // or, once standalone `super` compile is tweaked:
     // return getPropertyDescriptor(super, property).get();
    }
    
    get foo(): number { 
        return this.superGet('foo') + 1;
    }
}

var t = new Test3;
console.log(t.foo);
```

Combining the third option with @fdecampredon's suggestion:

> also generating something that throws an error in case of missing super setter and return undefined in case of missing super getter (like what traceur does https://github.com/google/traceur-compiler/blob/master/src/runtime/classes.js#L43-L60) seems pretty much standard and compliant with what future es6 engine will do. 

and @mwisnicki's suggestion:

> What if we lookup getters and setters only once, right before defining property ? For example like [this](https://gist.github.com/mwisnicki/4b32f48c3c34a6152014).

and tweaking @duncanmak's suggestion:

> Another possibility is to only allow it when emitting in ~~ES6~~ES5?

seems viable in the compiler (eliminating stringy / maintenance issues), and would cover the majority of use cases.

**Comment by vladima** [comment]
#5860 lifts the restriction for ES6, we still need a separate proposal for the downlevel emit

**Comment by jiaweihli** [comment]
@vladima I have a working solution (built on top of my [gist](https://gist.github.com/jiaweihli/807b56a75909a628fe7e), fixes a few bugs) for ES5, which we're running in production.  It uses 2 external dependencies, but these aren't strictly necessary and their functionality can be inlined.  Should I open a new issue?

**Comment by jiaweihli** [comment]
^ Bump @vladima

**Comment by pwnall** [comment]
@vladima Can you comment as to whether a proposal for the downlevel emit based on @jiaweihli's gist would have a chance of being accepted?

I don't think there is any solution that mitigates @ahejlsberg's first concern. However, in response to the second concern, TypeScript now has at least two more helpers to support [async/await](https://github.com/Microsoft/TypeScript/issues/1664). Also, this path will not descend into madness, as there is a finite set of features in ES6 that can be polyfilled in ES5.

**Comment by amcdnl** [comment]
Is there any open issues on this being merged downward to ES5? Moved from Babel to TS and had a lot of this going on, ended up doing something like:

```
  set sourcePosition(val: number) {
    // was working for babel
    // super.sourcePosition = val;

    // hack for ts :(
    Object.getOwnPropertyDescriptor(
      Object.getPrototypeOf(DirectedEdgeView.prototype), 'sourcePosition')
        .set.call(this, val);

    this.updateArrowPosition();
  }
```

**Comment by jpwiddy** [comment]
Any proposal to bring this to ES5?  It's very obscure, it doesn't work as expected. I'd really like this to be resolved.

**Comment by jiaweihli** [comment]
@vladima @RyanCavanaugh @ahejlsberg 

Any comments on my [proposal](https://github.com/Microsoft/TypeScript/issues/338#issuecomment-99180950)?

**Comment by jpwiddy** [comment]
@jiaweihli your proposal still doesn't work completely. It saves the value to the prototype, not the instance. So the super class's prototype will incorrectly have the value.

```
var getPropertyDescriptor = function(subject, name) {
  var pd = Object.getOwnPropertyDescriptor(subject, name);
  var proto = Object.getPrototypeOf(subject);
  while (pd === undefined && proto !== null) {
    pd = Object.getOwnPropertyDescriptor(proto, name);
    proto = Object.getPrototypeOf(proto);
  }
  return pd;
};

class Test {
    private _foo: any;
  superGet(property) {
    return getPropertyDescriptor(Object.getPrototypeOf(this.constructor.prototype), property).get();
    }
  superSet(property, value) {
    return getPropertyDescriptor(Object.getPrototypeOf(this.constructor.prototype), property).set(value);
    }

  constructor() { }

  set foo(val) {
      this._foo = val;
  }
  get foo(): number { return this._foo; }
}

class Test2 extends Test { }

class Test3 extends Test2 { }

class Test4 extends Test3 {
  get foo(): any {
      return this.superGet('foo');
  }
  set foo(val) {
      this.superSet('foo', val);
  }
}

var t = new Test4();
var t2 = new Test4();

t.foo = 123;
t2.foo = 321;

console.log(t.foo);  // undefined
console.log(t2.foo);  // undefined

console.log(Test4);  // has no foo value
console.log(Test);  // has no foo value 

```

**Comment by jiaweihli** [comment]
@jpwiddy 
Sorry for the slow update.

The proposal you tested is obsolete and incorrect, like you discovered 😛 .  Here's the new (and always up-to-date) proposal: https://gist.github.com/jiaweihli/807b56a75909a628fe7e

The new version [correctly works](https://runkit.com/5894d511a52a080014a2d293/58af4853084d0b0014a66379) with the example you gave.  This uses 2 external libraries at the moment for convenience, but these aren't strictly necessary and their functionality can be inlined.

**Comment by jpwiddy** [comment]
Hah no worries! Thanks for the updated gist! Looks better, I'd like to see the external dependencies alleviated in case this is pulled into TS. It'd make it much easier to merge in.

**Comment by Hotell** [comment]
This is a big deal breaker with Web Components and SkateJS. https://skatejs.gitbooks.io/skatejs/content/docs/api/Component.html#static-props

If user  want's to leverage mixins ( added in TS 2.2 ) this error is preventing to compile without errors which is very unfortunate:

```ts
interface Constructable<T> {
  readonly props: ComponentProps<any, any>;
  readonly observedAttributes: string[];
  readonly is: string;
  new ( ...args: any[] ): T,
}

export function Disabled<BC extends Constructable<{}>>( Base: BC ) {
  return class extends Base {
    static get props(): ComponentProps<any, DisabledProps> {
      return {
        ...super.props,
        disabled: prop.boolean( { attribute: true } )
      };
    }
    disabled?: boolean;
  };
}

class MyComponent extedns Component<{}>{
  static get props(){
    return {
      color: {}
    }
  }
  renderCallback(){
     const {disabled,color} = this;
     return (
       <button disabled={disabled} style={{color}}>Hello</button>
     )
   }
}
```

**Comment by kriszyp** [comment]
For the record, when we transitioned from Babel to TypeScript, I wrote this custom transformer + runtime patch. It's a bit of hack from a transformation perspective, but I believe it is semantically correct in behavior (and preserved our super accessor functionality as compiled with babel):
(before) transformer:
```
const ts = require('typescript')

function visitor(ctx, sf) {

  function visit(node) {
    if (ts.isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.EqualsToken && ts.isPropertyAccessExpression(node.left) && node.left.expression.kind === ts.SyntaxKind.SuperKeyword) {
      // super.property = value
      // transform to:
      // Object.superSet(_super.prototype, propertyName, this, value)
      return ts.createCall(
        ts.createPropertyAccess(
          ts.createIdentifier('Object'),
          'superSet'),
        [],
        [ts.createSuper(), ts.createLiteral(node.left.name), ts.createThis(), node.right])
    }
    if (ts.isPropertyAccessExpression(node) && node.expression.kind === ts.SyntaxKind.SuperKeyword &&
        // make sure we *don't* transform super.method() calls
        !(ts.isCallExpression(node.parent) && node.parent.expression === node)) {
      // super.property
      // transform to:
      // Object.superGet(_super.prototype, propertyName, this)
      return ts.createCall(
        ts.createPropertyAccess(
          ts.createIdentifier('Object'),
          'superGet'),
        [],
        [ts.createSuper(), ts.createLiteral(node.name), ts.createThis()])
    }
    return ts.visitEachChild(node, visit, ctx)
  }
  return visit
}

module.exports = function(/*opts?: Opts*/) {
  return (ctx) => {
    return (sf) => ts.visitNode(sf, visitor(ctx, sf))
  }
}
```
runtime patch:
```
Object.superGet = (superPrototype, name, instance) => {
  do {
    let descriptor = Object.getOwnPropertyDescriptor(superPrototype, name)
    if (descriptor) {
      return descriptor.get && descriptor.get.call(instance)
    }
  } while ((superPrototype = Object.getPrototypeOf(superPrototype)))
}

Object.superSet = (superPrototype, name, instance, value) => {
  do {
    let descriptor = Object.getOwnPropertyDescriptor(superPrototype, name)
    if (descriptor) {
      return descriptor.set && descriptor.set.call(instance, value)
    }
  } while ((superPrototype = Object.getPrototypeOf(superPrototype)))
  Object.defineProperty(instance, name, {
    value,
    configurable: true,
    enumerable: true,
    writable: true
  })
}
```
The runtime semantics require no knowledge of types or super class implementations. Obviously to put this in TypeScript proper, you would want that code to be generated, and/or in tslib, not patched on `Object`, though. As a before transformer, this relies on the fact that generating a `super` token will get transformed to `_super.prototype` in the later downlevel transformers. Naturally, would greatly prefer this in TypeScript itself, so maybe this will help.

**Comment by brianmhunt** [comment]
Thanks for reopening, @RyanCavanaugh — this is a problematic issue and readily worked around, per e.g. the patch by @kriszyp

**Comment by RyanCavanaugh** [comment]
Long story short - after five years of not having this and people surviving despite it, it seems like ES6 adoption is catching up fast enough that the narrow window of people who target ES5 (rather than 3 for maximum compat or 6 for smallest emit delta) is closing to the point where we don't want to accept a big chunk of helper emit at this point. Open to data otherwise but it just doesn't seem common enough to warrant a big change in the emitter.

**Comment by brianmhunt** [comment]
Just a data point: This is a blocker to using Typescript to compile Knockout.js version 4.  The depth of workaround needed is uncertain, but it feels problematic.

Cross-linking https://github.com/knockout/tko/issues/21

**Comment by thomashilzendegen** [comment]
I'm just curious, but when I put the following code into the playground, the output JS seems legit to me (and it works):

`class Foo {
    get bar() {
        return 'foo';
    }
}

class Bar extends Foo {
    get bar() {
        return super.bar + 'bar';
    }
}

const x = new Bar();
alert(x.bar);`

Does this mean, access to super's property works, but the compiler is still complaining?

**Comment by brianmhunt** [comment]
@ThomasdenH It's references to `this` that are broken.  Try:

```javascript
class Foo {
    constructor () { this.x = 'dee' }
    get bar() {
        return 'foo' + this.x;
    }
}

class Bar extends Foo {
    get bar() {
        return super.bar + 'bar';
    }
}

const foo = new Foo()
const bar = new Bar();
alert(`foo: ${foo.bar}, bar: ${bar.bar}`)
```

The output above is `foo: foodee, bar: barundefinedbar`, but when run in an ES6 engine (i.e. the correct outcome) it is `foo: foodee, bar: bardeebar`.

**Comment by thomashilzendegen** [comment]
@brianmhunt Brrr, I see. Good catch.

**Comment by SlurpTheo** [comment]
> The output above is `foo: foodee, bar: barundefinedbar`, but when run in an ES6 engine (i.e. the correct outcome) it is `foo: foodee, bar: bardeebar`.


@brianmhunt Little typo there:

The output above is `foo: foodee, bar: fooundefinedbar`, but when run in an ES6 engine (i.e. the correct outcome) it is `foo: foodee, bar: foodeebar`.

**Comment by brianmhunt** [comment]
@SlurpTheo Yes, you're correct; edited my comment.
